node/src/components/block_accumulator/tests.rs

```diff
@@ -61,3 +324,768 @@ fn upsert_acceptor() {
+            BTreeSet::from([peer_1, peer_2])
+        );
+        // The block should be marked complete in storage by now.
+        assert_eq!(
+            runner
+                .reactor()
+                .storage
+                .read_highest_complete_block()
+                .unwrap()
+                .unwrap()
+                .height(),
+            hot_block_1.block.height()
+        );
+    }
+
+    // Retrigger the event so the accumulator can update its hot block state.
+    {
+        runner
+            .process_injected_effects(|effect_builder| {
+                let event = super::Event::ExecutedBlock {
+                    hot_block: hot_block_1.clone(),
+                };
+                effect_builder
+                    .into_inner()
+                    .schedule(event, QueueKind::Validation)
+                    .ignore()
+            })
+            .await;
+        while runner.try_crank(&mut rng).await == TryCrankOutcome::NoEventsToProcess {
+            time::sleep(POLL_INTERVAL).await;
+        }
+    }
+
+    let older_block = generate_older_block(&mut rng, &block_1, 1);
+    // Register an older block.
+    {
+        let effect_builder = runner.effect_builder();
+        let reactor = runner.reactor_mut();
+
+        let block_accumulator = &mut reactor.block_accumulator;
+        let event = super::Event::ReceivedBlock {
+            block: Box::new(older_block.clone()),
+            sender: peer_1,
+        };
+        let effects = block_accumulator.handle_event(effect_builder, &mut rng, event);
+        assert!(effects.is_empty());
+        // This should have no effect on the accumulator since the block is
+        // older than the local tip.
+        assert!(!block_accumulator
+            .block_acceptors
+            .contains_key(older_block.hash()));
+    }
+
+    let older_block_signature = FinalitySignature::create(
+        *older_block.hash(),
+        older_block.header().era_id(),
+        &ALICE_SECRET_KEY,
+        ALICE_PUBLIC_KEY.clone(),
+    );
+    // Register a signature for an older block.
+    {
+        let effect_builder = runner.effect_builder();
+        let reactor = runner.reactor_mut();
+
+        let block_accumulator = &mut reactor.block_accumulator;
+        let event = super::Event::ReceivedFinalitySignature {
+            finality_signature: Box::new(older_block_signature),
+            sender: peer_2,
+        };
+        let effects = block_accumulator.handle_event(effect_builder, &mut rng, event);
+        assert!(effects.is_empty());
+        // This should *probably* have no effect on the accumulator since the
+        // block is older than the local tip, but right now it creates the
+        // acceptor if it's in the same era or newer than the local tip era.
+        // See comment in `BlockAccumulator::register_block`.
+        // assert!(!block_accumulator
+        //     .block_acceptors
+        //     .contains_key(older_block.hash()));
```

Comment:
This is registering a finality signature though rather than a block.

You could maybe provide a signature from an older era and check no new acceptor is created, and/or uncomment this but assert that the acceptor exists (with a comment about why it does).

Reply:
The problem scenario is this:
- we have an old block which, if received by the accumulator, would not have an acceptor created
- if you start by sending the signatures for said block, which creates the acceptor if the era is recent enough
- only after receiving the block we realize that this acceptor is useless
- it will eventually get purged

This is wasteful and arguably introduces unnecessary complexity. We could keep track of purged block hashes until their era stops being recent, but this would probably not be worth the hassle. Alternatively, we could just purge when we realize it's a stale block and live with it.

Reply:
We decided not to do anything about this. We accept this behavior to be the expected one.